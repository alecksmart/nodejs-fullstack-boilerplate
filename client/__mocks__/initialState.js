import { initialState as pointsGame } from '../src/reducers/pointsGame';

export default {
  pointsGame,
};
