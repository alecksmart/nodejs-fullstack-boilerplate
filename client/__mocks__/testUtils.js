import { shallow, mount, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';

configure({ adapter: new Adapter() });

export const mockStore = configureStore();

export const shallowWithStore = (component, store) => {
  const context = {
    store,
  };
  return shallow(component, { context });
};

export const mountWithStore = (component, store) => {
  const context = {
    store,
  };
  return mount(component, { context });
};
