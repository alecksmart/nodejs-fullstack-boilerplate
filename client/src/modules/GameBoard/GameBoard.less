.GameBoard {
  display: flex;
  height: 100vh;

  & h1 {
    background-color: darkslateblue;
    width: 100%;
    height: 40px;
    color: white;
    text-align: center;
    line-height: 40px;
    vertical-align: middle;
  }
}