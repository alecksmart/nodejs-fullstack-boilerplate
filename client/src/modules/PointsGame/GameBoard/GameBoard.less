.GameBoard {
  display: flex;
  height: 100vh;

  & h1 {
    background-color: darkslateblue;
    height: 40px;
    line-height: 40px;
    color: white;
    text-align: center;
    vertical-align: middle;
  }
}

@media all and (orientation:portrait) {
  .GameBoard {
    flex-direction: column;
  }
}

@media all and (orientation:landscape) {
  .GameBoard {
    flex-direction: row;
  }
}

