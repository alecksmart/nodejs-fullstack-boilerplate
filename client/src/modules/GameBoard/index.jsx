import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import GameField from '../GameField';
import Results from '../Results';
import Toolbar from '../Toolbar';

import { setNewGame, setCellClicked } from '../../actions/pointsGame';
import { makeMap } from '../../utils/pointsGame';

import './GameBoard.less';

class GameBoard extends PureComponent {
  constructor(props) {
    super(props);

    const pointsMap = makeMap();
    this.state = {
      pointsMap,
    };
  }

  onNewGame = () => {
    const { setNewGame } = this.props;
    const pointsMap = makeMap();
    this.setState({ pointsMap });
    setNewGame();
  }

  onPointSelected = (celldata) => {
    const { type } = celldata;
    const { setCellClicked } = this.props;
    setCellClicked({
      logEntry: { type },
    });
  }

  renderScores = () => (
    <ul className="Scores-Rules">
      <li>A: costs: 50, each 3-rd gives adds +50 bonus</li>
      <li>B: costs: 30, each 2-nd gives adds +30 bonus</li>
      <li>C: costs: 20</li>
      <li>B: costs: 15</li>
    </ul>
  );

  render() {
    const { pointsMap } = this.state;
    return (
      <div className="GameBoard">
        <GameField
          pointsMap={pointsMap}
          onPointSelected={this.onPointSelected}
        />
        <div className="Scores">
          <h1>How To Play</h1>
          <ol>
            <li>Click &quot;New Game&quot; below</li>
            <li>The field will randomly create 15 pills</li>
            <li>Open 10 pills of 15 trying to guess the best score</li>
            <li>
              Bonuses for pills types:
              {this.renderScores()}
            </li>
            <li>After opening 10 pills, the game is over, and the High Score count is updated</li>
            <li>Proceed to Step 1 to start over</li>
          </ol>
          <h1>SCORES</h1>
          <Results />
          <Toolbar onNewGame={this.onNewGame} />
        </div>
      </div>
    );
  }
}

GameBoard.propTypes = {
  setCellClicked: PropTypes.func.isRequired,
  setNewGame: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  isGameOver: state.pointsGame.isGameOver,
});

const mapDispatchToProps = {
  setNewGame,
  setCellClicked,
};

export default connect(mapStateToProps, mapDispatchToProps)(GameBoard);
