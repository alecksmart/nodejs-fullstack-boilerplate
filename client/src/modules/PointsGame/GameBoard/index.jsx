import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import GameField from '../GameField';
import Results from '../Results';
import Toolbar from '../Toolbar';

import { setNewGame, setCellClicked } from '../../../actions/pointsGame';
import { makeMap } from '../../../utils/pointsGame';

import './GameBoard.less';

export const GameBoard = ({ setNewGame, setCellClicked }) => {
  const [pointsMap, setPointsMap] = useState([]);
  useEffect(() => {
    setPointsMap(makeMap());
  }, []);

  const onNewGame = () => {
    setPointsMap(makeMap());
    setNewGame();
  };

  const onPointSelected = celldata => {
    const { type } = celldata;
    setCellClicked({
      logEntry: { type },
    });
  };

  const renderScores = () => (
    <ul className="Scores-Rules">
      <li>A: costs: 50, each 3-rd gives adds +50 bonus</li>
      <li>B: costs: 30, each 2-nd gives adds +30 bonus</li>
      <li>C: costs: 20</li>
      <li>B: costs: 15</li>
    </ul>
  );

  return (
    <div className="GameBoard">
      <GameField
        pointsMap={pointsMap}
        onPointSelected={onPointSelected}
      />
      <div className="Scores">
        <h1>Grab Points! Game</h1>
        <p>How To Play:</p>
        <ol>
          <li>Click &quot;New Game&quot; below</li>
          <li>The field will randomly create 15 pills</li>
          <li>Open 10 pills of 15 trying to guess the best score</li>
          <li>
            Bonuses for pills types:
            {renderScores()}
          </li>
          <li>After opening 10 pills, the game is over, and the High Score count is updated</li>
          <li>Proceed to Step 1 to start over</li>
        </ol>
        <h1>SCORES</h1>
        <Results />
        <Toolbar onNewGame={onNewGame} />
      </div>
    </div>
  );
};


GameBoard.propTypes = {
  setCellClicked: PropTypes.func.isRequired,
  setNewGame: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  setNewGame,
  setCellClicked,
};

export default connect(null, mapDispatchToProps)(GameBoard);
