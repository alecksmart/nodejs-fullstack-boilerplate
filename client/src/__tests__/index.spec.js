/* eslint-disable no-unused-vars */
/* eslint-disable import/no-unresolved */
import React from 'react';
import { Provider } from 'react-redux';
import { unmount } from 'enzyme';
import { mockStore, shallowWithStore, mountWithStore } from 'testUtils';
import initialState from 'initialState';
import GameBoard from '../modules/GameBoard/index';

let store;
let wrap;
let wrapper;
let context = initialState;
let useMount = true;

describe('>>> Client entry point with store', () => {
  beforeEach(() => {
    wrap = () => (useMount ? mountWithStore : shallowWithStore)(
      <Provider store={store}>
        <GameBoard />
      </Provider>,
      context,
    );
  });

  afterEach(() => {
    wrapper.unmount();
    useMount = true;
  });

  it('+++ should capture the snapshot of GameBoard', () => {
    useMount = false;
    store = mockStore(initialState);
    wrapper = wrap();
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it('+++ should show Game Over', () => {
    context = { ...context, pointsGame: { ...context.pointsGame, isGameOver: true } };
    store = mockStore(context);
    wrapper = wrap();
    const expectedText = 'Game Over, click the button above to try again...';
    expect(wrapper.find('.Toolbar-GameOver').text()).toEqual(expectedText);
  });

  it('+++ should create a new Game', () => {
    store = mockStore(context);
    wrapper = wrap();
    const mapBefore = wrapper.find('GameBoard').state().pointsMap;
    wrapper.find('GameBoard').instance().onNewGame();
    const mapAfter = wrapper.find('GameBoard').state().pointsMap;
    expect(mapBefore).not.toEqual(mapAfter);
  });
});
