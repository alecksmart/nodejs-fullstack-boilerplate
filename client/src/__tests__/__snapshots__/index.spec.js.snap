// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>> Client entry point with store +++ should capture the snapshot of GameBoard 1`] = `
"<ContextProvider value={{...}}>
  <Connect(GameBoard) />
</ContextProvider>"
`;
